### PR TITLE
Implement IANA time zone ID normalization

### LINKED ISSUE #4031: Support retrieval of normalize IANA names, including aliases
#4024 allows for accessing the canonical IANA identifier for a given time zone. However, it does not allow case normalization of non-canonical time zones. For example, Temporal requires that "asia/calcutta" be echoed back to the user as "Asia/Calcutta".

To implement this, add another data key that contains the normalized strings for all IANA names that are non-canonical. Sort them by their lowercase value, then look them up with a binary search.

Expose this as an API such as

```rust
pub struct AllInOneIanaBcp47Thingy {
    // should contain all 3 data keys
}

impl AllInOneIanaBcp47ThingyBorrowed {
    /// Returns the BCP-47 ID and the normalized IANA name of the input
    pub fn lookup_iana(&self, iana: &str) -> Option<(TimeZoneBcp47Id, String)> { ... }
}
```

### PR DESCRIPTION
**Pull request by sffc**
Fixes #4031

### CODE REVIEW

**Comment by sffc** [comment]
Blocked on #4562

**Comment by sffc** [comment]
> Blocked on #4562

Not really blocked on this, but we should resolve #4562 before shipping 1.5

File: provider/datagen/src/tests/make_testdata.rs
Comment (robertbastian) [inline provider/datagen/src/tests/make_testdata.rs:1]: please sync

File: components/timezone/src/provider.rs [hunk 1]
```diff
@@ -42,6 +42,7 @@ const _: () = {
     icu_timezone_data::make_provider!(Baked);
     icu_timezone_data::impl_time_zone_bcp47_to_iana_v1!(Baked);
     icu_timezone_data::impl_time_zone_iana_to_bcp47_v1!(Baked);
+    icu_timezone_data::impl_time_zone_iana_to_bcp47_v2!(Baked);
```
Comment (robertbastian) [inline components/timezone/src/provider.rs:45]: if the v2 struct supports all v1 functionality, we should not include baked data for both

File: components/timezone/src/provider/names.rs [hunk 2]
```diff
@@ -52,6 +53,54 @@ pub struct IanaToBcp47MapV1<'data> {
...
+))]
+#[cfg_attr(
+    feature = "datagen", 
+    derive(serde::Serialize, databake::Bake),
+    databake(path = icu_timezone::provider::names),
+)]
+#[cfg_attr(feature = "serde", derive(serde::Deserialize))]
+pub struct IanaToBcp47MapV2<'data> {
+    /// A map from IANA time zone identifiers to indexes of BCP-47 time zone identifiers.
+    /// The lowest bit is 1 if the name is canonical and 0 if it is not canonical.
```
Comment (robertbastian) [inline components/timezone/src/provider/names.rs:79]: I don't understand what this means. Please phrase this in a way that does not require knowing ZeroTrie internals.

File: components/timezone/src/provider/names.rs [hunk 3]
```diff
@@ -52,6 +53,54 @@ pub struct IanaToBcp47MapV1<'data> {
...
+    /// The IANA identifiers are normal-case.
+    #[cfg_attr(feature = "serde", serde(borrow))]
+    pub map: ZeroAsciiIgnoreCaseTrie<ZeroVec<'data, u8>>,
+    /// A sorted list of BCP-47 time zone identifiers.
+    #[cfg_attr(feature = "serde", serde(borrow))]
+    // Note: this is 9739B as ZeroVec<TinyStr8> and 9335B as VarZeroVec<str>
```
Comment (robertbastian) [inline components/timezone/src/provider/names.rs:85]: And how much as `ZeroVec<'data, TimeZoneBcp47Id>`?

File: components/timezone/src/provider/names.rs (hunk 2 above)
Comment (sffc) [reply components/timezone/src/provider/names.rs:79]: Addressed in #4718

File: components/timezone/src/provider/names.rs (hunk 3 above)
Comment (sffc) [reply components/timezone/src/provider/names.rs:85]: Addressed in #4718

**Comment by sffc** [comment]
Putting this back in draft so I can rework the API a bit more.

**Comment by sffc** [comment]
Replaced by https://github.com/unicode-org/icu4x/pull/4774

**Pull request closed without merging**